pkg/platform/runtime/setup/setup.go

```diff
@@ -222,41 +222,24 @@ func (s *Setup) installArtifacts(buildResult *model.BuildResult, artifacts artif
 	}
 	// Artifacts are installed in two stages
 	// - The first stage runs concurrently in MaxConcurrency worker threads (download, unpacking, relocation)
-	// - The second stage moves all files into its final destination is running in a single thread to avoid file conflicts
-
-	// secondStageCh is the channel over which the final installation routines are started
-	secondStageCh := make(chan func() error)
+	// - The second stage moves all files into its final destination is running in a single thread (using the mainthread library) to avoid file conflicts
+	var artfErrs []error
 
-	// schedule the first stage
-	var firstStageError error
-	go func() {
-		defer close(secondStageCh)
+	// schedule the first stage, binding mainthread library to this thread
+	mainthread.Run(func() {
+		var err error
 		if buildResult.BuildReady {
-			firstStageError = s.installFromBuildResult(buildResult, artifacts, secondStageCh)
+			err = s.installFromBuildResult(buildResult, artifacts, artfErrs)
 		} else {
-			firstStageError = s.installFromBuildLog(buildResult, artifacts, secondStageCh)
+			err = s.installFromBuildLog(buildResult, artifacts, artfErrs)
 		}
-	}()
 
-	// run the move operations in the main thread
-	var artfErrs []error
-	for f := range secondStageCh {
-		err := f()
 		if err != nil {
-			artfErrs = append(artfErrs, errs.Wrap(err, "Error during execution of second stage function"))
-		}
-	}
-
-	// add err message from the first stage
-	if firstStageError != nil {
-		// merge with artifact setup error if possible
-		var ase *ArtifactSetupErrors
-		if errors.As(firstStageError, &ase) {
-			artfErrs = append(artfErrs, ase.Errors()...)
-		} else {
-			artfErrs = append(artfErrs, firstStageError)
+			mainthread.Call(func() {
+				artfErrs = append(artfErrs, err)
+			})
```

Comment:
I don't think we should do this for updating a slice of errors. For that we should just use channels. Reserve `mainthread` for logic that is otherwise complicated to facilitate with channels. eg. moving a bunch of files from different artifacts without needing to combine them all into a single channel when that's not how the data flows.

Reply:
I pushed a solution using channels, it required another go routine though...